### PR TITLE
Managed texture loading

### PR DESCRIPTION
**Pull request by jprochazk**
Part of https://github.com/emilk/egui/issues/3291.

- [x] Add types from the proposal
- [x] Add loader implementations from proposal to `egui_extras`
- [x] Update usage in examples
- [x] Documentation

### CODE REVIEW

File: crates/epaint/src/image.rs [hunk 1]
```diff
@@ -11,7 +12,7 @@ use crate::{textures::TextureOptions, Color32};
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
 pub enum ImageData {
     /// RGBA image.
-    Color(ColorImage),
+    Color(Arc<ColorImage>),
```
Comment (emilk) [inline crates/epaint/src/image.rs:15]: Or add the `Arc` to inside `ColorImage` 🤷

File: crates/epaint/src/image.rs [hunk 2]
```diff
@@ -226,6 +227,12 @@ impl std::ops::IndexMut<(usize, usize)> for ColorImage {
 impl From<ColorImage> for ImageData {
     #[inline(always)]
     fn from(image: ColorImage) -> Self {
+        Self::Color(Arc::new(image))
+    }
+}
+
+impl From<Arc<ColorImage>> for ImageData {
+    fn from(image: Arc<ColorImage>) -> Self {
```
Comment (emilk) [inline crates/epaint/src/image.rs:235]:
```suggestion
    #[inline]
    fn from(image: Arc<ColorImage>) -> Self {
```

**Review by emilk — CHANGES_REQUESTED** [review]
Really awesome!

**Comment by jprochazk** [comment]
> Image loading can be slow.

@emilk is it slow eonugh that we should also make `ImageCrateLoader` load images on a separate thread?

**Comment by emilk** [comment]
> > Image loading can be slow.
> 
> @emilk is it slow eonugh that we should also make `ImageCrateLoader` load images on a separate thread?

The preference here actually depend on the application.

If you only have a few images, you may prefer a slow first frame than the flicker that comes from loading things in the background (also goes for loading the bytes from disk). But if you have hundreds of images, then you definitely want to load them in background threads. But spawning one thread for each image (or two - one for the loading from disk, one for the parsing of the image file) is not ideal, so we should also consider adding `rayon` as an opt-in dependency for that.

I'm not sure how a user would want to configure this. Probably with a flag in `egui::Options` that's then passed in to the loaders. But let's punt on that for now - let's start with always using background threads, either with `std::thread` or `rayon`, if that feature is enabled. That will yield the most responsive applications, and is pretty standard these days (that's what people is used to on the web, and in electron, etc)

File: crates/egui_extras/src/loaders.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/// Install the default set of loaders:
+/// - `file` loader on non-Wasm targets
+/// - `http` loader with the `ehttp` feature
+/// - `image` loader with the `image` feature
+///   - the supported set of image formats is configured by enabling `image` crate features.
+/// - `svg` loader with the `svg` feature
+///
+/// The `file` and `http` loaders are bytes loaders, they do not know how to turn those
+/// bytes into an image. `svg` is for loading `.svg` files, and `image` can load any
+/// other image format enabled on the `image` crate.
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/egui_extras/src/loaders.rs:12]: Documentation is tricky! Hopefully I made it clearer here

File: crates/egui_extras/src/loaders/ehttp_loader.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
+use egui::{
+    ahash::HashMap,
+    load::{BytesLoadResult, BytesLoader, BytesPoll, LoadError},
+    mutex::Mutex,
+};
+use std::{sync::Arc, task::Poll};
+
+#[derive(Default)]
+pub struct EhttpLoader {
+    cache: Arc<Mutex<HashMap<String, Poll<Result<Arc<[u8]>, String>>>>>,
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/egui_extras/src/loaders/ehttp_loader.rs:10]: I'm not sure if those two are the same. `BytesLoadResult` contains `BytesPoll` in the `Ok` variant, so this change would make it something like `Poll<Result<Poll>>`, which I believe is not what we want here.

**Comment by jprochazk** [comment]
> let's start with always using background threads, either with std::thread or rayon, if that feature is enabled

Didn't think this through properly, but spawning a background thread for anything other than `FileLoader` would not work for Wasm. I say let's punt on it completely for now and figure out later how we'll make it non-blocking on Wasm, or if we do that at all.

**Comment by jprochazk** [comment]
One thing that isn't done here in any way is constructing an `Image2` from non-static bytes, e.g. `egui::ColorImage::example()`. Not exactly sure how that should work. Should the `load_include_bytes` function and `IncludeBytesLoader` just allow non-`'static` lifetimes?

**Comment by emilk** [comment]
> One thing that isn't done here in any way is constructing an `Image2` from non-static bytes, e.g. `egui::ColorImage::example()`. Not exactly sure how that should work. Should loading from bytes just allow non-`'static` lifetimes?

I think so - the `IncludeBytesLoader` still copies the bytes to an `Arc<[u8]>` anyways, so why not.

A potential optimization is to replace `Arc<[u8]>` with `enum Bytes { Static(&'static [u8]), Shared(Arc<[u8]>) }` to avoid both the copies and the extra memory use.

**Review by emilk — APPROVED** [review]
Exciting!

File: crates/egui/src/load.rs [hunk 5]
```diff
@@ -274,17 +314,27 @@ pub trait TextureLoader {
...
-        self.cache.lock().entry(uri).or_insert_with(|| bytes.into());
+impl DefaultBytesLoader {
+    pub(crate) fn insert_static(&self, uri: &'static str, bytes: &'static [u8]) {
+        self.cache
+            .lock()
+            .entry(uri)
+            .or_insert_with(|| Bytes::Static(bytes));
+    }
+
+    pub(crate) fn insert_shared(&self, uri: &'static str, bytes: impl Into<Arc<[u8]>>) {
```
Comment (emilk) [inline crates/egui/src/load.rs:329]:
```suggestion
    pub(crate) fn insert_static(&self, uri: &'static str, bytes: impl Into<Bytes>) {
        self.cache
            .lock()
            .entry(uri)
            .or_insert_with(|| bytes.into());
    }
```

**Comment by jprochazk** [comment]
I'm not totally sure about the API I came up with for `Bytes`, but we can always improve it later.

File: crates/egui/src/load.rs (hunk 5 above)
Comment (jprochazk) [reply crates/egui/src/load.rs:329]: There's a tiny optimization here: `Bytes` requires `Arc<[u8]>` to already be constructed, `impl Into<Arc<[u8]>>` doesn't. So here the user would have to copy their bytes into a `Arc<[u8]>` first if the data is not `'static`, where as with this split, we only copy them if there isn't already an entry for the given `uri`.

Is it worth keeping?

File: crates/epaint/src/texture_handle.rs [hunk 6]
```diff
@@ -95,6 +95,11 @@ impl TextureHandle {
         crate::Vec2::new(w as f32, h as f32)
     }
 
+    /// `width x height x bytes_per_pixel`
+    pub fn byte_size(&self) -> usize {
+        self.tex_mngr.read().meta(self.id).unwrap().bytes_used()
```
[the remark this answers is not in the document]
Comment (jprochazk) [reply crates/epaint/src/texture_handle.rs:100]: The `size` method also unwraps, so I just copied that. I'll change both to not unwrap.

**Pull request merged**